src/main/java/org/openrewrite/java/migrate/maven/UpdateMavenProjectPropertyJavaVersion.java

```diff
@@ -97,28 +96,30 @@ public Xml.Document visitDocument(Xml.Document document, ExecutionContext ctx) {
 
                 // Otherwise override remote parent's properties locally
                 MavenResolutionResult mrr = getResolutionResult();
-                Map<String, String> currentProperties = mrr.getPom().getRequested().getProperties();
-                for (String property : JAVA_VERSION_PROPERTIES) {
-                    if (currentProperties.containsKey(property) || !propertiesExplicitlyReferenced.contains(property)) {
-                        continue;
+                boolean foundProperty = false;
+                for (Map.Entry<String, String> property : mrr.getPom().getProperties().entrySet()) {
+                    if (JAVA_VERSION_PROPERTIES.contains(property.getKey())) {
+                        foundProperty = true;
+                        try {
+                            if (Float.parseFloat(property.getValue()) < version) {
+                                d = (Xml.Document) new AddProperty(property.getKey(), String.valueOf(version), null, false)
+                                        .getVisitor()
+                                        .visitNonNull(d, ctx);
+                            }
+                        } catch (NumberFormatException ex) {
+                            // either an expression or something else, don't touch
+                        }
                     }
-                    d = (Xml.Document) new AddProperty(property, String.valueOf(version), null, false)
-                            .getVisitor()
-                            .visitNonNull(d, ctx);
                 }
 
                 // When none of the relevant properties are explicitly configured Maven defaults to Java 8
                 // The release option was added in 9
                 // If no properties have yet been updated then set release explicitly
-                if (version >= 9 &&
-                    !compilerPluginConfiguredExplicitly &&
-                    currentProperties.keySet()
-                        .stream()
-                        .noneMatch(JAVA_VERSION_PROPERTIES::contains)) {
+                if (!foundProperty && version >= 9 && !compilerPluginConfiguredExplicitly) {
                     d = (Xml.Document) new AddProperty("maven.compiler.release", String.valueOf(version), null, false)
                             .getVisitor()
                             .visitNonNull(d, ctx);
-                    HashMap<String, String> updatedProps = new HashMap<>(currentProperties);
+                    HashMap<String, String> updatedProps = new HashMap<>(mrr.getPom().getRequested().getProperties());
                     updatedProps.put("maven.compiler.release", version.toString());
                     mrr = mrr.withPom(mrr.getPom().withRequested(mrr.getPom().getRequested().withProperties(updatedProps)));
```

Comment:
I don’t understand why this is needed, It’s not done for the other `AddProperty` executions.

Reply:
This was to prevent a second cycle of the loop: after adding the property to the `mrr`, it would prevent entering this `if` again.